### DISCUSSION TITLE
Consider relaxing requirement that types/namespaces must come after top level statements.

### DISCUSSION THREAD

**Discussion by CyrusNajmabadi** [dotnet/csharplang#5780] [Language Ideas]
User feedback on top level statements showed the following as a pattern they commonly use:

```c#
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

using Corporate.Interfacing;

namespace Corporate.Process
{
    static class Program
    {
        internal class Options : Corporate.Interfacing.ArgumentHandler
        {
            [Corporate.Interfacing.CommandLineParameter(Description = "The input folder", LongIdentifier = "inFolder", ShortIdentifier = "i")]
            internal string? Path { get; set; }
            [Corporate.Interfacing.CommandLineParameter(Description = "The output folder", LongIdentifier = "output", ShortIdentifier = "o")]
            internal string? OutDir { get; set; }
            [Corporate.Interfacing.CommandLineParameter(Description = "The download folder", LongIdentifier = "downloads", ShortIdentifier = "d")]
            internal string? DownloadDir { get; set; }
        }
        private static Options options = new Options();

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        public static int Main(string[] args)
        {
            if (!options.HandleArgs(args) || (options.Path == null || options.OutDir == null || options.DownloadDir == null) || args.Length < 3)
            {
                Console.Error.WriteLine(" !! Missing arguments !! ");
                options.PrintUsage();
                return -1;
            }
            Corporate.Process.Processor.OutputDir = options.OutDir;
            Corporate.Process.Processor.DownloadDir = options.DownloadDir;
            Corporate.Process.Processor.LoadFile(options.Path);
            Corporate.Process.Processor.ProcessFile();
            return 0;
        }
    }
}
```

While this can pretty trivially be converted to top level statements, the code now looks like the following:

```c#
Options options = new Options();

if (!options.HandleArgs(args) || (options.Path == null || options.OutDir == null || options.DownloadDir == null) || args.Length < 3)
{
    Console.Error.WriteLine(" !! Missing arguments !! ");
    options.PrintUsage();
    return -1;
}

Corporate.Process.Processor.OutputDir = options.OutDir;
Corporate.Process.Processor.DownloadDir = options.DownloadDir;
Corporate.Process.Processor.LoadFile(options.Path);
Corporate.Process.Processor.ProcessFile();
return 0;

class Options : Corporate.Interfacing.ArgumentHandler
{
    [Corporate.Interfacing.CommandLineParameter(Description = "The input folder", LongIdentifier = "inFolder", ShortIdentifier = "i")]
    internal string? Path { get; set; }
    [Corporate.Interfacing.CommandLineParameter(Description = "The output folder", LongIdentifier = "output", ShortIdentifier = "o")]
    internal string? OutDir { get; set; }
    [Corporate.Interfacing.CommandLineParameter(Description = "The download folder", LongIdentifier = "downloads", ShortIdentifier = "d")]
    internal string? DownloadDir { get; set; }
}
```

While this does have the cleanliness benefits that come with not needing hte scaffolding from before, it has a downside in that the pattern of having the necessary auxiliary types come first before they are used no longer holds in this code.  Instead, things have flipped to 'procedural code first, helper code next'. 

It does not seem like it would be strange or undesirable to relax things here to allow types/namespaces (or potentially just types), to precede the procedural code of the top level statements.  That would then appear as:

```c#
class Options : Corporate.Interfacing.ArgumentHandler
{
    [Corporate.Interfacing.CommandLineParameter(Description = "The input folder", LongIdentifier = "inFolder", ShortIdentifier = "i")]
    internal string? Path { get; set; }
    [Corporate.Interfacing.CommandLineParameter(Description = "The output folder", LongIdentifier = "output", ShortIdentifier = "o")]
    internal string? OutDir { get; set; }
    [Corporate.Interfacing.CommandLineParameter(Description = "The download folder", LongIdentifier = "downloads", ShortIdentifier = "d")]
    internal string? DownloadDir { get; set; }
}

/// <summary>
/// The main entry point for the application.
/// </summary>

Options options = new Options();

if (!options.HandleArgs(args) || (options.Path == null || options.OutDir == null || options.DownloadDir == null) || args.Length < 3)
{
    Console.Error.WriteLine(" !! Missing arguments !! ");
    options.PrintUsage();
    return -1;
}

Corporate.Process.Processor.OutputDir = options.OutDir;
Corporate.Process.Processor.DownloadDir = options.DownloadDir;
Corporate.Process.Processor.LoadFile(options.Path);
Corporate.Process.Processor.ProcessFile();
return 0;
```

This seems equally readable while matching stylistic preference of types-first.   

This would *not* include the ability to define types *in situ* with statements.

**Comment by CyrusNajmabadi** [comment]
Ok.  I'll Champion this work.

**Comment by DavidArno** [comment]
To me, the appearance of:
```cs
/// <summary>
/// The main entry point for the application.
/// </summary>
```
encapsulates the downside to this proposal: suddenly it's unclear where the top level code begins and a comment has to be used as an "it's here" marker. One has to now potentially scroll past other things first in order to find that code. By mandating that the code must come first, that situation is avoided: the reader knows the order they will encounter things and comments as markers to where things are aren't then required.

Whilst the current opinionated approach may not to be everyone's liking (it's not the way I'd normally lay out the code), it has the big advantage of being simple and consistent. So in my view, this is an example of a feature suggestion that starts out at -100 and rapidly *decreases*.

**Reply by CyrusNajmabadi** [reply]
> and a comment has to be used as an "it's here" marker.

It doesn't have to be. I used that becuse that was just part of the users original code. They had that on their Main method as well.

**Comment by orthoxerox** [comment]
I've never liked the Pascal style of having the helper code come first.

**Comment by V0ldek** [comment]
I'd like to offer the counter view to some opinions that this isn't as clear or somehow harder to understand.

This'd make single-file programs much more natural to write. When I write a script in F#, or in Rust, or in Python, I structure the script based on its logic. Which means that if there's some logic in phase 2 for which I want to declare a type then I want to put that type between phase 1 and 2, not in an arbitrary clump, be it at the top ot the bottom of the file.

It'd also make example programs easier to explain naturally, line-by-line, introducing types when they're needed, instead of, again, in a clump.

**Comment by MadL1me** [comment]
I would love to see this. I'm personally, when trying ideas or making examples in console application, make types at first, and then write code, and 95% of the time I'm required to move all my types at bottom of the file, which is quite annoying
